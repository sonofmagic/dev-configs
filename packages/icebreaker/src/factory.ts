import type {
  ConfigNames,
  TypedFlatConfigItem,
} from '@antfu/eslint-config'
import type { FlatConfigComposer } from 'eslint-flat-config-utils'
import { isPackageExists } from 'local-pkg'
import { defu } from 'defu'
import { antfu, interopDefault } from './antfu'
import type { UserConfigItem, UserDefinedOptions } from './types'

export function icebreaker(
  options: UserDefinedOptions = {},
  ...userConfigs: UserConfigItem[]
): FlatConfigComposer<TypedFlatConfigItem, ConfigNames> {
  const {
    tailwindcss: enableTailwindcss = isPackageExists('tailwindcss'),
    mdx: enableMDX,
    a11y: enableA11y,
    ...opts
  } = defu<UserDefinedOptions, UserDefinedOptions[]>(options, {
    formatters: true,
  })

  const presets: UserConfigItem[] = [
    {
      rules: {
        'curly': ['error', 'all'],
        'no-console': ['warn'],
        'ts/prefer-ts-expect-error': 'off',
        'ts/ban-ts-comment': 'off',
        'vue/attribute-hyphenation': 'off',
        // 问题在于 auto fix 的时候，会直接 remove 整个 import ，而我们想让用户自己去 remove
        // 'unused-imports/no-unused-imports': 'error',
        // https://typescript-eslint.io/rules/no-unused-vars/
        'no-unused-vars': 'off',
        'ts/no-unused-vars': 'error',
      },
    },
  ]

  // https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/335
  if (enableTailwindcss) {
    presets.push(
      interopDefault(
        // @ts-ignore
        import('eslint-plugin-tailwindcss'),
      ).then((tailwind) => {
        return tailwind.configs['flat/recommended']
      }),
    )
    presets.push({
      rules: {
        'tailwindcss/no-custom-classname': 'off',
      },
    })
  }

  if (enableMDX) {
    presets.push(interopDefault(import('eslint-plugin-mdx')).then((mdx) => {
      return [
        {
          ...mdx.flat,
          processor: mdx.createRemarkProcessor({
            lintCodeBlocks: true,
            languageMapper: {},
          }),
        },
        {
          ...mdx.flatCodeBlocks,
          rules: {
            ...mdx.flatCodeBlocks.rules,
          },
        },
      ]
    }))
  }

  if (enableA11y) {
    if (opts.vue) {
      presets.push(
        interopDefault(
          // @ts-ignore
          import('eslint-plugin-vuejs-accessibility'),
        ).then((pluginVueA11y) => {
          return pluginVueA11y.configs['flat/recommended']
        }),
      )
    }

    if (opts.react) {
      presets.push(
        interopDefault(
          // @ts-ignore
          import('eslint-plugin-jsx-a11y'),
        ).then((jsxA11y) => {
          return jsxA11y.flatConfigs.recommended
        }),
      )
    }
  }

  return antfu(opts, ...presets, ...userConfigs)
}
