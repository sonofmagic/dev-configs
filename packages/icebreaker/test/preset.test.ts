import { getRestConfigAndPresets } from '@/factory'

describe('presets', () => {
  it('c', () => {
    expect(getRestConfigAndPresets()).toMatchSnapshot()
  })
  it('c0', () => {
    expect(getRestConfigAndPresets({ a11y: true })).toMatchSnapshot()
  })
  it('c1', () => {
    expect(getRestConfigAndPresets({ vue: true })).toMatchSnapshot()
  })
  it('c2', () => {
    expect(getRestConfigAndPresets({ vue: {
      vueVersion: 3,
    } })).toMatchSnapshot()
  })
  it('c3', () => {
    expect(getRestConfigAndPresets({
      vue: {
        vueVersion: 2,
      },
    })).toMatchSnapshot()
  })
})
