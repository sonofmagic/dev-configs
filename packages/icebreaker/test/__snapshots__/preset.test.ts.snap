// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`presets > c 1`] = `
[
  {
    "formatters": true,
    "javascript": {
      "overrides": {
        "curly": [
          "error",
          "all",
        ],
        "no-console": [
          "warn",
        ],
        "no-unused-vars": "off",
      },
    },
    "typescript": {
      "overrides": {
        "ts/ban-ts-comment": "off",
        "ts/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true,
          },
        ],
        "ts/no-unused-vars": [
          "error",
          {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "ignoreRestSiblings": true,
            "varsIgnorePattern": "^_",
          },
        ],
        "ts/no-use-before-define": "warn",
        "ts/prefer-ts-expect-error": "off",
      },
    },
  },
  {
    "rules": {
      "unused-imports/no-unused-vars": "off",
    },
  },
]
`;

exports[`presets > c0 1`] = `
[
  {
    "a11y": true,
    "formatters": true,
    "javascript": {
      "overrides": {
        "curly": [
          "error",
          "all",
        ],
        "no-console": [
          "warn",
        ],
        "no-unused-vars": "off",
      },
    },
    "typescript": {
      "overrides": {
        "ts/ban-ts-comment": "off",
        "ts/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true,
          },
        ],
        "ts/no-unused-vars": [
          "error",
          {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "ignoreRestSiblings": true,
            "varsIgnorePattern": "^_",
          },
        ],
        "ts/no-use-before-define": "warn",
        "ts/prefer-ts-expect-error": "off",
      },
    },
  },
  {
    "rules": {
      "unused-imports/no-unused-vars": "off",
    },
  },
]
`;

exports[`presets > c1 1`] = `
[
  {
    "formatters": true,
    "javascript": {
      "overrides": {
        "curly": [
          "error",
          "all",
        ],
        "no-console": [
          "warn",
        ],
        "no-unused-vars": "off",
      },
    },
    "typescript": {
      "overrides": {
        "ts/ban-ts-comment": "off",
        "ts/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true,
          },
        ],
        "ts/no-unused-vars": [
          "error",
          {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "ignoreRestSiblings": true,
            "varsIgnorePattern": "^_",
          },
        ],
        "ts/no-use-before-define": "warn",
        "ts/prefer-ts-expect-error": "off",
      },
    },
    "vue": {
      "overrides": {
        "vue/attribute-hyphenation": "off",
        "vue/custom-event-name-casing": "off",
        "vue/no-mutating-props": "warn",
        "vue/no-unused-refs": "warn",
        "vue/no-useless-v-bind": [
          "error",
          {
            "ignoreIncludesComment": false,
            "ignoreStringEscape": true,
          },
        ],
        "vue/no-v-for-template-key": "off",
        "vue/no-v-for-template-key-on-child": "error",
        "vue/v-on-event-hyphenation": "off",
      },
    },
  },
  {
    "rules": {
      "unused-imports/no-unused-vars": "off",
    },
  },
]
`;

exports[`presets > c2 1`] = `
[
  {
    "formatters": true,
    "javascript": {
      "overrides": {
        "curly": [
          "error",
          "all",
        ],
        "no-console": [
          "warn",
        ],
        "no-unused-vars": "off",
      },
    },
    "typescript": {
      "overrides": {
        "ts/ban-ts-comment": "off",
        "ts/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true,
          },
        ],
        "ts/no-unused-vars": [
          "error",
          {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "ignoreRestSiblings": true,
            "varsIgnorePattern": "^_",
          },
        ],
        "ts/no-use-before-define": "warn",
        "ts/prefer-ts-expect-error": "off",
      },
    },
    "vue": {
      "overrides": {
        "vue/attribute-hyphenation": "off",
        "vue/custom-event-name-casing": "off",
        "vue/no-mutating-props": "warn",
        "vue/no-unused-refs": "warn",
        "vue/no-useless-v-bind": [
          "error",
          {
            "ignoreIncludesComment": false,
            "ignoreStringEscape": true,
          },
        ],
        "vue/no-v-for-template-key": "off",
        "vue/no-v-for-template-key-on-child": "error",
        "vue/v-on-event-hyphenation": "off",
      },
      "vueVersion": 3,
    },
  },
  {
    "rules": {
      "unused-imports/no-unused-vars": "off",
    },
  },
]
`;

exports[`presets > c3 1`] = `
[
  {
    "formatters": true,
    "javascript": {
      "overrides": {
        "curly": [
          "error",
          "all",
        ],
        "no-console": [
          "warn",
        ],
        "no-unused-vars": "off",
      },
    },
    "typescript": {
      "overrides": {
        "ts/ban-ts-comment": "off",
        "ts/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true,
          },
        ],
        "ts/no-unused-vars": [
          "error",
          {
            "args": "all",
            "argsIgnorePattern": "^_",
            "caughtErrors": "all",
            "caughtErrorsIgnorePattern": "^_",
            "destructuredArrayIgnorePattern": "^_",
            "ignoreRestSiblings": true,
            "varsIgnorePattern": "^_",
          },
        ],
        "ts/no-use-before-define": "warn",
        "ts/prefer-ts-expect-error": "off",
      },
    },
    "vue": {
      "overrides": {
        "vue/attribute-hyphenation": "off",
        "vue/custom-event-name-casing": "off",
        "vue/no-deprecated-v-bind-sync": "off",
        "vue/no-mutating-props": "warn",
        "vue/no-unused-refs": "warn",
        "vue/no-useless-v-bind": [
          "error",
          {
            "ignoreIncludesComment": false,
            "ignoreStringEscape": true,
          },
        ],
        "vue/no-v-for-template-key": "error",
        "vue/no-v-for-template-key-on-child": "off",
        "vue/v-on-event-hyphenation": "off",
      },
      "vueVersion": 2,
    },
  },
  {
    "rules": {
      "unused-imports/no-unused-vars": "off",
    },
  },
]
`;
